File: src/main/java/dev/retrotv/crypto/owe/hash/crc/CRC32.java
package dev.retrotv.crypto.owe.hash.crc;

import dev.retrotv.crypto.owe.hash.Checksum;
import dev.retrotv.crypto.owe.hash.Hash;
import dev.retrotv.crypto.owe.hash.PasswordWithSalt;
import dev.retrotv.utils.EncodeUtil;
import dev.retrotv.utils.MessageDigestEncodeUtil;

import static dev.retrotv.enums.HashAlgorithm.CRC32;

/**
 * CRC-32 알고리즘으로 암호화 하기 위한 {@link Checksum}, {@link PasswordWithSalt} 인터페이스의 구현체 입니다.
 *
 * @author  yjj8353
 * @since   1.8
 */
public class CRC32 extends Hash {

    @Override
    public String hash(byte[] data) {

        // 앞에 0이 패딩되는 부분을 무시하고 뒤의 8자리만 잘라낸다
        return EncodeUtil.binaryToHex(MessageDigestEncodeUtil.encode(CRC32, data))
                         .substring(8);
    }
}

File: src/main/kotlin/dev/retrotv/crypto/owe/hash/crc/CRC32.kt
package dev.retrotv.crypto.owe.hash.crc

import dev.retrotv.crypto.owe.hash.Hash
import dev.retrotv.data.utils.ByteUtils
import dev.retrotv.enums.Algorithm.Hash.CRC32
import dev.retrotv.utils.hashing

/**
 * CRC-32 알고리즘으로 암호화 하기 위한 [Hash] 추상 클래스의 구현체 입니다.
 *
 * @author  yjj8353
 * @since   1.0.0
 */
class CRC32 : Hash() {

    init {
        this.algorithm = CRC32
    }

    override fun hash(data: ByteArray): String {
        return ByteUtils.toHexString(hashing(this.algorithm, data)).substring(8)
    }

    override fun upgradeEncoding(encodedPassword: String?): Boolean {
        log.debug("파일 해싱 이외의 용도로 사용중일 경우, 알고리즘 업그레이드를 권장합니다.")
        return true
    }
}
